### PR TITLE
threefry_2x32

### PR DESCRIPTION
**Pull request by wozeparrot**
Closes: #2418

### CODE REVIEW

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -156,9 +156,27 @@ def empty(*shape, **kwargs):
...
   def manual_seed(seed=0): Tensor._seed = seed
 
+  _rng_counter: int = 0
   @staticmethod
   def rand(*shape, **kwargs):
-    return Tensor._loadop(LoadOps.CUSTOM, prod((shape:=argfix(*shape))), arg=custom_random, **kwargs).reshape(shape)
+    num = prod((shape:=argfix(*shape)))
+    if (odd_counts := num % 2): num += 1
+    counts = Tensor.arange(num, dtype=dtypes.uint32) + Tensor._rng_counter
+    Tensor._rng_counter += num
```
Comment (blueridanus) [inline tinygrad/tensor.py:165]: **note**: this works with early realizes but breaks if lazy

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -156,9 +156,27 @@ def empty(*shape, **kwargs):
   @staticmethod
   def manual_seed(seed=0): Tensor._seed = seed
 
+  _rng_counter: int = 0
```
Comment (gabelerner) [inline tinygrad/tensor.py:159]: from what i read, this will need to be device specific. 

how i was planning to do it in my PR was:
`_rng_counter = defaultdict(int)` and then hash device like `device_id = int(hashlib.sha256('CPU'.encode()).hexdigest(), 16) & 0xFFFFFFFF` to get the current counter for the device

also i think below the `0x0` in `ks` should be the `device_id` 

that way the key/counter contain:
- seed
- device id
- device specific counter

which should allow parallel use without getting the same numbers twice

**Comment by wozeparrot** [comment]
> I realized something about XOR. Isn't it just add on a bool Tensor? (though numpy seems to behave differently)
> 
> Never mind, add behaves like OR
> 
> ```
> #include <stdio.h>
> #include <stdbool.h>
> int main() {
>   bool a = true;
>   bool b = true;
>   bool c = a+b;
>   printf("%d\n", c);
> }
> ```
> 
> Oh, xor is SUB! (tested in C)
> 
> ```
> 0 - 0 = 0
> 1 - 0 = 1
> 0 - 1 = 1
> 1 - 1 = 0
> ```
> 
> So it should be possible to bitcast this to bool and subtract them. Would just be in the `__xor__` definition on Tensor, but that way we don't need a new op.

Is it guaranteed that bool has the same bitwidth as uint32 on all backends?

**Comment by wozeparrot** [comment]
hmm, torch doesn't support uint32 as a dtype

**Comment by geohot** [comment]
I'm fine with merging this with XOR, the bool thing has a whole bunch of prereqs we don't have (size changing bitcast for example)

**Comment by wozeparrot** [comment]
yea, gonna put xor in a seperate pr

File: tinygrad/tensor.py [hunk 3]
```diff
@@ -153,12 +153,31 @@ def empty(*shape, **kwargs):
...
+  def manual_seed(seed=0): Tensor._seed, Tensor._rng_counter = seed, Tensor([0], dtype=dtypes.uint32, requires_grad=False)
 
   @staticmethod
   def rand(*shape, **kwargs):
-    return Tensor._loadop(LoadOps.CUSTOM, prod((shape:=argfix(*shape))), arg=custom_random, **kwargs).reshape(shape)
+    if Tensor._rng_counter is None: Tensor._rng_counter = Tensor([0], dtype=dtypes.uint32, requires_grad=False)
+    num = prod((shape:=argfix(*shape)))
+    if num == 0: return Tensor.zeros(shape, **kwargs)
+    counts = (Tensor.arange(num, dtype=dtypes.uint32, requires_grad=None if kwargs.get("requires_grad") is None else kwargs.get("requires_grad")) + Tensor._rng_counter).realize()
+    if (odd_counts := num % 2): counts = counts.cat(Tensor([0], dtype=dtypes.uint32, requires_grad=False))
```
Comment (geohot) [inline tinygrad/tensor.py:166]: Just padding here should work

Comment (wozeparrot) [reply tinygrad/tensor.py:166]: its actually easier just to do `if (odd_counts := num % 2): num += 1` but then it doesn't match the jax impl.

**Comment by geohot** [comment]
I'm also fine with hardcoding a few (regenerable by jax) references instead of installing jax. This is pretty typical for crypto stuff

**Comment by geohot** [comment]
I'm fine with keeping the old rand for torch since torch doesn't support uint32

File: extra/optimization/test_beam_search.py [hunk 4]
```diff
@@ -14,38 +14,38 @@ def tearDown(self):
...
 
   def test_variable_ast_beam(self):
-    a = Tensor.rand(3, 3).reshape((Variable("a", 1, 10).bind(3), 3))
+    a = Tensor.empty(3, 3).reshape((Variable("a", 1, 10).bind(3), 3))
     a = (a+1).realize()
 
   def test_big_prime_number(self):
-    a = Tensor.rand(367, 367)
-    b = Tensor.rand(367, 367)
+    a = Tensor.empty(367, 367)
```
Comment (geohot) [inline extra/optimization/test_beam_search.py:21]: These tests can't be empty, they need real numbers to compare with numpy

File: tinygrad/tensor.py [hunk 5]
```diff
@@ -152,11 +152,29 @@ def _loadop(op, shape, device:Optional[str]=None, dtype:Optional[DType]=None, ar
...
+    rotations = [[13, 15, 26, 6], [17, 29, 16, 24]]
+    ks = [0x0, Tensor._seed ^ 0x0 ^ 0x1BD11BDA, Tensor._seed]
+
+    x = [(c := counts.chunk(2))[0] + ks[-1], c[1] + ks[0]]
+    for i in range(5):
+      for r in rotations[i % 2]: x[0], x[1] = (x0 := x[0] + x[1]), x0 ^ ((x[1] * (2 ** r)) + (x[1] / (2 ** (32 - r))))
+      x = [(x[0] + ks[i % 3]), (x[1] + ks[(i + 1) % 3] + i + 1)]
+    out = (x[0].cat(x[1])[:num].cast(dtypes.float32).realize() / (2 ** 32 - 1)).reshape(shape).cast(dtypes.default_float if dtype is None else dtype)
+    out.requires_grad = kwargs.get("requires_grad")
+    return out
```
Comment (geohot) [inline tinygrad/tensor.py:177]: Would you not have to change the tests if you added a contiguous here? I think it's fine for rand to not fuse with other ops.

Comment (wozeparrot) [reply tinygrad/tensor.py]: Most of the test changes aren't actually caused by rand fusing, just that rand itself is slower. the beam search test I don't really want to change, but without changing it it is really slow as its beam searching the rand as well.

I do think that having the contiguous is better anyways.

File: tinygrad/features/search.py [hunk 6]
```diff
@@ -37,7 +37,7 @@ def _time_program(ast:LazyOp, rdev:Compiled, lib:bytes, global_size, local_size,
   tms = []
   for _ in range(cnt):
     if clear_l2:
-      with Context(DEBUG=0): Tensor.rand(1024,1024).realize()
+      with Context(DEBUG=0): Tensor.ones(1024,1024).realize()
```
Comment (chenyuxyz) [inline tinygrad/features/search.py:40]: this needs to be a real buffer to clear l2 cache

**Comment by geohot** [comment]
What's blocking this?

**Comment by wozeparrot** [comment]
I'm not sure what happened with metal, started failing pretty recently, and its hard for me to reproduce as I don't have a mac.

Also on CPU this takes enormous amounts of ram.

Still running into that `__repr__` issue with beam search since the asts are really deep.

**Comment by geohot** [comment]
torch is going to be removed soon, so this should be doable.

**Comment by wozeparrot** [comment]
Still optional? or just full commit?

**Comment by geohot** [comment]
Can we get this merged optionally to start?

**Comment by wozeparrot** [comment]
Currently I have a flag to turn it off, should I flip that so that its opt-in?

**Comment by geohot** [comment]
Yea, I think opt in is good so there's not THREEFRY=0 everywhere in CI

**Comment by wozeparrot** [comment]
CI being slower makes sense, threefry generates quite a few more kernels running the tests.

**Comment by chenyuxyz** [comment]
I think it's fine to skip the randomness half test for threefry. That does not work for bfloat16 either and we need to have a different fix

**Comment by chenyuxyz** [comment]
can you sync to master? I added some bf16 rand stuff that I can clean up after this is merged

**Comment by geohot** [comment]
Congrats, bounty is yours
